ckan/controllers/user.py

```diff
@@ -80,7 +80,8 @@ def _setup_template_variables(self, context, data_dict):
         try:
             user_dict = get_action('user_show')(context, data_dict)
         except NotFound:
-            abort(404, _('User not found'))
+            h.flash_error('Authorize to see this page')
+            h.redirect_to(controller='user', action='login')
```

Comment:
I think `NotAuthenticated` is a good idea (just like HTTP which distinguishes between 401 and 403), but it would have to be implemented properly: The whole authentication work-flow around `ckan.logic.check_access` would have to be updated accordingly.